Typing is failing tests on Python 3.9.0

This shows up on a wheel build at https://tinyurl.com/3475ussr
```

    @pytest.mark.parametrize("name,func", [
        ("__init__", lambda n: n),
        ("__init__", lambda n: _GenericAlias(np.ndarray, Any)),
        ("__init__", lambda n: _GenericAlias(np.ndarray, (Any,))),
        ("__init__", lambda n: _GenericAlias(np.ndarray, (Any, Any))),
        ("__init__", lambda n: _GenericAlias(np.ndarray, T1)),
        ("__init__", lambda n: _GenericAlias(np.ndarray, (T1,))),
        ("__init__", lambda n: _GenericAlias(np.ndarray, (T1, T2))),
        ("__origin__", lambda n: n.__origin__),
        ("__args__", lambda n: n.__args__),
        ("__parameters__", lambda n: n.__parameters__),
        ("__reduce__", lambda n: n.__reduce__()[1:]),
        ("__reduce_ex__", lambda n: n.__reduce_ex__(1)[1:]),
        ("__mro_entries__", lambda n: n.__mro_entries__([object])),
        ("__hash__", lambda n: hash(n)),
        ("__repr__", lambda n: repr(n)),
        ("__getitem__", lambda n: n[np.float64]),
        ("__getitem__", lambda n: n[ScalarType][np.float64]),
        ("__getitem__", lambda n: n[Union[np.int64, ScalarType]][np.float64]),
        ("__getitem__", lambda n: n[Union[T1, T2]][np.float32, np.float64]),
        ("__eq__", lambda n: n == n),
        ("__ne__", lambda n: n != np.ndarray),
        ("__dir__", lambda n: dir(n)),
        ("__call__", lambda n: n((1,), np.int64, BUFFER)),
        ("__call__", lambda n: n(shape=(1,), dtype=np.int64, buffer=BUFFER)),
        ("subclassing", lambda n: _get_subclass_mro(n)),
        ("pickle", lambda n: n == pickle.loads(pickle.dumps(n))),
        ("__weakref__", lambda n: n == weakref.ref(n)()),
    ])
    def test_pass(self, name: str, func: FuncType) -> None:
        """Compare `types.GenericAlias` with its numpy-based backport.
    
        Checker whether ``func`` runs as intended and that both `GenericAlias`
        and `_GenericAlias` return the same result.
    
        """
        value = func(NDArray)
    
        if sys.version_info >= (3, 9):
>           value_ref = func(NDArray_ref)

func       = <function TestGenericAlias.<lambda> at 0x7f25d5a98700>
name       = '__weakref__'
self       = <numpy.typing.tests.test_generic_alias.TestGenericAlias object at 0x7f25aaf80790>
value      = True

/venv/lib/python3.9/site-packages/numpy/typing/tests/test_generic_alias.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

n = numpy.ndarray[typing.Any, numpy.dtype[+ScalarType]]

>       ("__weakref__", lambda n: n == weakref.ref(n)()),
    ])
E   TypeError: cannot create weak reference to 'types.GenericAlias' object

n          = numpy.ndarray[typing.Any, numpy.dtype[+ScalarType]]

/venv/lib/python3.9/site-packages/numpy/typing/tests/test_generic_alias.py:69: TypeError
```

--- a/numpy/typing/tests/test_generic_alias.py
+++ b/numpy/typing/tests/test_generic_alias.py
@@ -66,7 +66,6 @@ class TestGenericAlias:
         ("__call__", lambda n: n(shape=(1,), dtype=np.int64, buffer=BUFFER)),
         ("subclassing", lambda n: _get_subclass_mro(n)),
         ("pickle", lambda n: n == pickle.loads(pickle.dumps(n))),
-        ("__weakref__", lambda n: n == weakref.ref(n)()),
     ])
     def test_pass(self, name: str, func: FuncType) -> None:
         """Compare `types.GenericAlias` with its numpy-based backport.
@@ -79,6 +78,14 @@ class TestGenericAlias:
 
         if sys.version_info >= (3, 9):
             value_ref = func(NDArray_ref)
+            assert value == value_ref
+
+    def test_weakref(self) -> None:
+        """Test ``__weakref__``."""
+        value = weakref.ref(NDArray)()
+
+        if sys.version_info >= (3, 9, 1):  # xref bpo-42332
+            value_ref = weakref.ref(NDArray_ref)()
             assert value == value_ref
 
     @pytest.mark.parametrize("name", GETATTR_NAMES)

